oci.go

```diff
@@ -0,0 +1,113 @@
+//
+// Copyright (c) 2018 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// OCI config file
+const (
+	ociConfigFile     string      = "config.json"
+	ociConfigFileMode os.FileMode = 0444
+)
+
+// changeToBundlePath changes the cwd to the bundle path defined in the OCI spec
+func changeToBundlePath(spec *specs.Spec) (string, error) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		return cwd, err
+	}
+
+	if spec == nil || spec.Root == nil || spec.Root.Path == "" {
+		return cwd, errors.New("Invalid OCI spec")
+	}
+
+	rootfsPath := spec.Root.Path
+	if !filepath.IsAbs(rootfsPath) {
+		rootfsPath = filepath.Join(cwd, rootfsPath)
+	}
+
+	bundlePath := filepath.Dir(rootfsPath)
+
+	err = os.Chdir(bundlePath)
+	return cwd, err
+}
+
+// writeSpecToFile writes the container's OCI spec to 'config.json'
+// in the bundle path as expected by the OCI specification.
+func writeSpecToFile(spec *specs.Spec) error {
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	configPath := path.Join(cwd, ociConfigFile)
+	f, err := os.OpenFile(configPath, os.O_RDWR|os.O_CREATE, ociConfigFileMode)
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+	err = json.NewEncoder(f).Encode(spec)
+	return err
+}
+
+func isValidHook(file os.FileInfo) (bool, error) {
+	if file.IsDir() {
+		return false, fmt.Errorf("is a directory")
+	}
+
+	mode := file.Mode()
+	if (mode & os.ModeSymlink) != 0 {
+		return false, fmt.Errorf("is a symbolic link")
+	}
+
+	perm := mode & os.ModePerm
+	if (perm & 0111) == 0 {
+		return false, fmt.Errorf("is not executable")
+	}
+
+	return true, nil
+}
+
+// findHooks searches guestHookPath for any OCI hooks for a given hookType
+func findHooks(guestHookPath, hookType string) (hooksFound []specs.Hook) {
+	hooksPath := path.Join(guestHookPath, hookType)
+
+	files, err := ioutil.ReadDir(hooksPath)
+	if err != nil {
+		agentLog.WithError(err).Infof("Skipping hook type %s", hookType)
+		return
+	}
+
+	for _, file := range files {
+		name := file.Name()
+		if ok, err := isValidHook(file); !ok {
+			agentLog.WithError(err).Warnf("Skipping hook %s", name)
+			continue
+		}
+
+		agentLog.Infof("Adding hook %s of type %s", name, hookType)
```

Comment:
```
agentLog.WithFields(logrus.Fields{
    "oci-hook-name" : name,
    "oci-hook-type" : hookType,
}).Info("Adding hook")
```

Reply:
I think this is a nice example of why I don't like this logging model, but I will do the change :)